### PR TITLE
Doc/3199 monitoring adr

### LINKED ISSUE #3199: Performance Monitoring ADR
**Description:**
_Provide a brief background and justification for this issue_
_OFA & UX team: Please include bullets capturing the value to STT end users, related guidance, and/or talking points that should be communicated to STTs in release notes_

**Acceptance Criteria:** 

- [ ] Architecture Decision Record for Application Health Monitoring with expected audience of OCIO
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] _Task 1_
- [ ] _Task 2_
- [ ] _Task 3_
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- _Note 1_
- _Note 2_
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- _Doc 1_
- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- _Open Question 1_
- _Open Question 2_

### PR DESCRIPTION
**Pull request by andrew-jameson**
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #3199

Please also see #3206

### CODE REVIEW

**Review by raftmsohani — APPROVED** [review]
LGTM!

File: docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# 22. Monitoring Application Health and Performance
+
+Date: 2024-09-30
+
+## Status
+
+Pending
```
Comment (andrew-jameson) [inline docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md:7]: This will remain Pending until both @ADPennington and OFA's OCIO sign-off. Reminder, this is not live but just a proposal as prerequisite for #3042 .

File: docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+Additionally, we have experienced critical blocking issues related to our updates to both Elasticsearch (ES) and PostgreSQL, which have compounded the need for more proactive alerting and load-testing in lower environments. Without timely notifications, we risk delays in addressing failures that could escalate into more significant problems. 
+
+
+## Decision
+We will build out a suite of tools in accordance with industry best practices to monitor our applications. Implementing a comprehensive monitoring and alerting ecosystem will not only help in identifying errors in real-time but also enable us to establish benchmarks based on historical data. This approach will foster a more proactive response strategy, ensuring that potential issues are mitigated before they impact our users or that system owners are aware of issues that have impacted users.
```
Comment (ADPennington) [inline docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md]:
```suggestion
We will build out a suite of tools in accordance with industry best practices to monitor our applications. Implementing a comprehensive monitoring and alerting ecosystem will not only help in identifying errors in real-time but also enable us to establish benchmarks based on historical data. This approach will foster a more proactive response strategy, ensuring that potential issues are mitigated before they impact our users or that system owners and system admins are aware of issues that have impacted users.
```

File: docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+![Detail exceptions](../images/sentry/3.%20detail%20about%20exception.png)
+
+<p style="text-align:center; margin:0; padding:0;">Full stack trace of the exceptions</p>
+
+![Full stack trace of the exceptions](../images/sentry/4.%20full%20stack%20trace%20of%20the%20exceptions.png)
+
+
+Performance monitoring in Sentry can greatly enhance backend application by providing real-time insights into how TANF app is performing. It tracks various metrics such as response time, database queries, and external API calls, with which we can identify performance bottlenecks to our backend app.
+
+A unique ability of Sentry is that it can link performance issues and group them together. This gives us the ability to attack and resolve more critical issues with highest impact first. Not only it can detect issues with web transactions, it also detects problems with DB queries as well as function regressions (if the duration of function has increased)
```
Comment (ADPennington) [inline docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md]:
```suggestion
A unique ability of Sentry is that it can link performance issues and group them together. This gives us the ability to attack and resolve more critical issues with highest impact first. Not only can it detect issues with web transactions, it also detects problems with DB queries as well as function regressions (if the duration of function has increased)
```

File: docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+### Why Prometheus-Loki-Grafana
+
+Grafana shall provide a visualization dashboard for these various tools which will collect and aggregate performance metrics, system logs, and allow deeper analysis for all aspects of our systems: frontend, proxies, backend, databases, and even networking. Additionally, the development team will seek to hone a proactive alerting system for out-of-threshold issues and errors for improved visibility of system issues.
+
+The storing of system logs will allow more expedient troubleshooting and debugging that is currently out of reach with Cloud.gov's existing Kibana interface. The ability to find and correlate log events is critical to technical analysis of faults, performance degradation, and system's overall health.
```
Comment (ADPennington) [inline docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md]:
```suggestion
The storing of system logs will allow more expedient troubleshooting and debugging that is currently out of reach with Cloud.gov's existing Kibana interface for logging. The ability to find and correlate log events is critical to technical analysis of faults, performance degradation, and system's overall health.
```

File: docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+### Why Prometheus-Loki-Grafana
+
+Grafana shall provide a visualization dashboard for these various tools which will collect and aggregate performance metrics, system logs, and allow deeper analysis for all aspects of our systems: frontend, proxies, backend, databases, and even networking. Additionally, the development team will seek to hone a proactive alerting system for out-of-threshold issues and errors for improved visibility of system issues.
+
+The storing of system logs will allow more expedient troubleshooting and debugging that is currently out of reach with Cloud.gov's existing Kibana interface for logging. The ability to find and correlate log events is critical to technical analysis of faults, performance degradation, and system's overall health.
+
+By having our monitoring ecosystem take in performance metrics, we will garner performance metrics over time as opposed to simply a live snapshot as is currently provided. This will allow to spotting of anomolous or out-of-bounds behaviors such as out of memory, high memory, cpu spikes, and disk thrashing.
```
Comment (ADPennington) [inline docs/Technical-Documentation/Architecture-Decision-Record/022-app-monitoring.md]:
```suggestion
By having our monitoring ecosystem take in performance metrics, we will garner performance metrics over time as opposed to simply a live snapshot as is currently provided. This will allow spotting of anomolous or out-of-bounds behaviors such as out of memory, high memory, cpu spikes, and disk thrashing.
```

**Review by jtimpe — APPROVED** [review]

**Review by ADPennington — COMMENTED** [review]
@andrew-jameson @lfrohlich @ttran-hub -- below is my feedback related to this ADR, most of these are geared toward trying to anticipate questions from our product and security teams: 

--- 

#### diagram-related

- recommend adding the ATO boundary to make it more clear that this solution will be hosted outside of it
- where are the other tools (PLG and associated) in this diagram?
- Does the triangle represent load balancers? 

#### cost-related
- Can we lay out the monthly costs associated with hosting Sentry in Cloud.gov compared to using Sentry's SaaS offering?
- Do we have a preliminary sense of what might increase these monthly costs over time (on the Sentry SaaS side)? What additional expenses (e.g., scaling costs) might be incurred as part of this solution?

#### security-related

- Does self-hosting mean that the other tools (PLG stack and associated) will be hosted in Cloud.gov? If so, what are the associated costs, and how do these compare to other hosting options? If not, where will this be hosted, and do we have visibility into its security compliance standards (e.g., encryption, access controls)?
- What security and compliance standards does the Sentry SaaS adhere to (e.g., SOC 2, FedRAMP, HIPAA)?
- Is the data stored in Sentry’s SaaS encrypted in transit and at rest? If so, what encryption standards are used?
- Will any PII or other sensitive data be stored outside of our ATO boundary? If so, please describe how it will be protected.

#### access control-related

- Can you describe the types of information to be exchanged between our system’s boundary and Sentry’s SaaS, including what logs, metrics, or other data will be sent or accessed? And what info will _not_ be exchanged?
- How will sys admins access this information, and will it require accessing data outside of our system’s boundary diagram?
- If information flows back into our boundary diagram, how are we ensuring that this process doesn’t create a new vector for potential attacks or data leaks?
- What access controls and monitoring mechanisms will be in-place for these external systems to detect and prevent unauthorized access?

**Review by elipe17 — APPROVED** [review]
LGTM

**Comment by ADPennington** [comment]
> @andrew-jameson @lfrohlich @ttran-hub -- below is my feedback related to this ADR, most of these are geared toward trying to anticipate questions from our product and security teams:
> 
> #### diagram-related
> * recommend adding the ATO boundary to make it more clear that this solution will be hosted outside of it

Andrew to denote this in diagram

> * where are the other tools (PLG and associated) in this diagram?

 under clamav in diagram 
plg


> * Does the triangle represent load balancers?

triangle represents the proxies

> #### cost-related
> * Can we lay out the monthly costs associated with hosting Sentry in Cloud.gov compared to using Sentry's SaaS offering?

$26/mo for Sentry SaaS vs $130/mo * 8GB-10GB if we hosted Sentry in our cloud.gov environment

reference: https://sentry.io/pricing/ 

> * Do we have a preliminary sense of what might increase these monthly costs over time (on the Sentry SaaS side)? What additional expenses (e.g., scaling costs) might be incurred as part of this solution?

could go up to $80/mo (Sentry pricing Business tier) if we need custom dashboards. it's too early to decide this. 


> #### security-related
> * Does self-hosting mean that the other tools (PLG stack and associated) will be hosted in Cloud.gov? If so, what are the associated costs, and how do these compare to other hosting options? If not, where will this be hosted, and do we have visibility into its security compliance standards (e.g., encryption, access controls)?

PLG to be hosted in cloud.gov requires about 10GB (*$130/mo), but not sure just yet. 


> * What security and compliance standards does the Sentry SaaS adhere to (e.g., SOC 2, FedRAMP, HIPAA)?

SOC 2 (reference: https://sentry.io/security/#third-party-audit)

> * Is the data stored in Sentry’s SaaS encrypted in transit and at rest? If so, what encryption standards are used?
https://sentry.io/security/#data-security-and-privacy

- PII scrubbing (reference: https://docs.sentry.io/security-legal-pii/scrubbing/)
- encryption at rest and in transit (reference: https://sentry.io/security/#data-encryption)

> * Will any PII or other sensitive data be stored outside of our ATO boundary? If so, please describe how it will be protected.

- possible in the logs, but we will be conducting an audit prior to moving this into prod to mitigate the potential for this (related to #3224 and #2562)

> #### access control-related
> * Can you describe the types of information to be exchanged between our system’s boundary and Sentry’s SaaS, including what logs, metrics, or other data will be sent or accessed? And what info will _not_ be exchanged?

- log information (e.g. exceptions, metrics, etc. no sensitive PII) 
- to get to the dashboard, we'd log into sentry (similar to how we would go to cloud.gov --reference: https://sentry.io/security/#single-sign-on); possible that we could integrate with AMS SSO ; dev does not have access to prod so would need another way to review sentry logs to support system admins

> * How will sys admins access this information, and will it require accessing data outside of our system’s boundary diagram?

see above

> * If information flows back into our boundary diagram, how are we ensuring that this process doesn’t create a new vector for potential attacks or data leaks?

mostly outbound communication (not sure about inbound yet - TBD spike)

> * What access controls and monitoring mechanisms will be in-place for these external systems to detect and prevent unauthorized access?

must go through nginx to reach our backend. if sentry compromised... incident response (https://sentry.io/security/#corporate-security) we may need our own IR plan. 


@lfrohlich  TDP IPT (external) is next Wednesday -- dev plans to present this. cc: @andrew-jameson @vlasse86 @ttran-hub

**Comment by lfrohlich** [comment]
The additional info above is helpful but @andrew-jameson  could you summarize what the expected cost would be, or the range of costs with all components considered?

**Comment by andrew-jameson** [comment]
> The additional info above is helpful but @andrew-jameson could you summarize what the expected cost would be, or the range of costs with all components considered?

@lfrohlich On low end, total would be $806/mo.

Per @elipe17, the range for PLG is approx 6-10GB of RAM total to cover all 3 spaces; above quote is for 6GB. Sentry would be $26/mo flat fee.

> We need to actually calculate memory requirements for PLG. My assumption is Loki: 2-4GB, Prometheus: 2-4GB ([calculator](https://www.robustperception.io/how-much-ram-does-prometheus-2-x-need-for-cardinality-and-ingestion/)), Grafana: 1GB, 3 PG exporters: 3 * 24MB, 6 Backend Promtails: 6 * 64MB.

**Comment by elipe17** [comment]
> > The additional info above is helpful but @andrew-jameson could you summarize what the expected cost would be, or the range of costs with all components considered?
> 
> @lfrohlich On low end, total would be $806/mo.
> 
> Per @elipe17, the range for PLG is approx 6-10GB of RAM total to cover all 3 spaces; above quote is for 6GB. Sentry would be $26/mo flat fee.
> 
> > We need to actually calculate memory requirements for PLG. My assumption is Loki: 2-4GB, Prometheus: 2-4GB ([calculator](https://www.robustperception.io/how-much-ram-does-prometheus-2-x-need-for-cardinality-and-ingestion/)), Grafana: 1GB, 3 PG exporters: 3 * 24MB, 6 Backend Promtails: 6 * 64MB.

Please see my updated memory requirements [here](https://github.com/raft-tech/TANF-app/pull/3192#issuecomment-2435525037).

**Comment by andrew-jameson** [comment]
> Please see my updated memory requirements [here](https://github.com/raft-tech/TANF-app/pull/3192#issuecomment-2435525037).

Thank you @elipe17! New low-end cost would be $416/mo @lfrohlich

**Comment by ADPennington** [comment]
> > Please see my updated memory requirements [here](https://github.com/raft-tech/TANF-app/pull/3192#issuecomment-2435525037).
> 
> Thank you @elipe17! New low-end cost would be $416/mo @lfrohlich

low end: 3GB x $130 (RAM cost per GB)+ $26 (Sentry) = $416/mo
high end (doubling GB estimate for budgeting purposes): 6GB x $130 + $26 = $806/mo

@lfrohlich

**Review by ADPennington — APPROVED** [review]
approved, pending feedback from OCIO. if additional changes are needed we can track those in a new ticket. cc: @lfrohlich
